src/core/imap.ts

```diff
@@ -0,0 +1,117 @@
+import * as Imap from 'imap';
+import {ImapConfig} from '../model/imapConfig.model';
+import {Query} from '../model/Query.model';
+
+export class IMAP {
+    private imap = null;
+    private monitorId = null;
+    private query: Query = null;
+    private delayTime = 1000 * 5;
+
+    constructor(config: ImapConfig) {
+        this.imap = new Imap({
+            ...config,
+            authTimeout: 10000,
+            connTimeout: 30000,
+            tlsOptions: {
+                rejectUnauthorized: false,
+            },
+
+        });
+    }
+
+    init() {
+        return new Promise((resolve, reject) => {
+            this.imap.connect();
+
+            this.imap.once('ready', ()=> {
+                console.log('%c---------------------  SUCCESSFUL IMAP CONNECTION  --------------------', 'color: Green');
+                resolve(this.imap);
+            });
+
+            this.imap.once('error', function(err: Error) {
+                reject(err);
+            });
+        });
+    }
+
+    setQuery(query: Query) {
+        this.query = query;
+    }
+
+    openBox(mailBox, readOnly = true) {
+        return new Promise((resolve, reject) => {
+            this.imap.openBox(mailBox, readOnly, (err, box) => {
+                if (err) {
+                    reject(err);
+                } else {
+                    resolve(box);
+                }
+            });
+        });
+    }
+
+    search(criteria) {
+        return new Promise((resolve, reject) => {
+            this.imap.search(criteria, (err, messages) => {
+                if (err) {
+                    reject(err);
+                } else {
+                    resolve(messages);
+                }
+            });
+        });
+    }
+
+    state() {
+        return new Promise(async (resolve, reject) => {
+            if (!navigator.onLine) {
+                reject(new Error('No internet Connection'));
+            } else if (!this.imap) {
+                reject(new Error('Please Re-login'));
+            } else if (this.imap.state === 'authenticated') {
+                resolve('authenticated');
+            } else {
+                try {
+                    await this.init();
+                    resolve('Reconnected successfully');
+                } catch (err) {
+                    reject(err);
+                }
+            }
+        });
+    }
+
+    monitor() {
+        let mailBox = null;
+        let criteria = null;
+
+        this.monitorId = setInterval(async () => {
+            //
+            if (this.query) {
+                ({mailBox, criteria} = this.query);
+
+                try {
+                    // Check if the connection is still stable.
+                    await this.state();
+
+                    await this.openBox(mailBox);
```

Comment:
Same here, this returns `box` that you never use.

Reply:
Done :)

Reply:
how does the change look like I cannot find anything in the latest commit? 
It could be that I am just too much of coding novice to see it :)

Reply:
> how does the change look like I cannot find anything in the latest commit? It could be that I am just too much of coding novice to see it :)

Briefly, Dealing with operations in imap is asynchronous because we are waiting for a response from the mail service provider with approval or rejection, so this is not an instantaneous process.

That is why you will notice the word 'await' beside the function.

When this data is available from the email service provider, we use the 'resolve' and can also return the box or message from the email provider depending on what the function does.

But we don't need this data now, so what I [did](https://github.com/joplin/plugin-email/pull/4/commits/e8784fec62dec940231c443f06a80289ae8b5780) was just `resolve()` without value.

```
Before:
	resolve(box);
After:
	resolve();
```
If there is anything unclear, please tell me.

Reply:
thx I found the change in the code, eventually, in https://github.com/joplin/plugin-email/pull/4/commits/e8784fec62dec940231c443f06a80289ae8b5780, thx for pushing me in the right direction

thx for the nice explanation :), can you show me where resolve() is defined?

Reply:
You’re welcome,

> can you show me where resolve() 

###  Inside promise function 
```
let promise = new Promise(function(`resolve`, reject) {

});
```
Resolve and reject are callback functions, and calling the resolve function if the job is finished successfully.

If you are interested in the promise, this [website](https://javascript.info/promise-basics) is very nice.